### PR TITLE
adding str.format() support

### LINKED ISSUE #90: format is not implemented
"str {0}".format(something) fails to run properly.

### LINKED ISSUE #321: .format() is not working
.format is a standard function so I don't understand why it doesn't work: example:

print "The answer is {}".format(42)

does not work.

ERROR = AttributeError: 'str' object has no attribute 'format'

### PR DESCRIPTION
**Pull request by gerbal**
str.format is almost 100% supported. Fixes #90, fixes #321 

Some issues remain in repr support and I have no idea how to handle kwargs, any advice is welcome.

I'm sure I have overlooked some edge cases of various combinations of formatting specifications.

### CODE REVIEW

**Comment by bnmnetp** [comment]
Thanks for the PR. 

If you haven't looked at it already, the HACKING.rst file has some information on doing kwargs.

Brad

**Comment by bnmnetp** [comment]
One problem is that **kwargs does not work properly in Skulpt.  This is a separate issue that should not keep this PR from getting accepted.  

```python
coord = {'latitude': '37.24N', 'longitude': '-115.81W'}

def x(latitude=1, longitude=1):
   print latitude, longitude

x()
x(**coord)
```

fails on line 7.

**Comment by gerbal** [comment]
Good to know about **kwargs. I disabled the test that cause Travis failures on **kwargs.

I'm still not sure why sk.builtin.repr() is failing in the remaining unittest. Any insight?

File: src/formatting.js [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            precision = parseInt(precision, 10);
+        }
+
+        formatFormat = function(value){
+            var r;
+            if(conversion === undefined || conversion === ""){
+                return value;
+            }
+            else if( conversion == "r"){
+                r = Sk.builtin.repr(value);
```
Comment (bnmnetp) [inline src/formatting.js]: The problem may be that by the time you are here, value is a raw javascript string.  It is no longer a python object.  ``repr`` does not deal with javascript objects, hence it returns unknown.

Comment (gerbal) [inline src/formatting.js]: Thanks! With that cleared up, all test are passing.

**Comment by bnmnetp** [comment]
I added the following test to test_strformat.py, however it failed.

```
    def test_width(self):
        self.assertEqual('         2,2',"{0:10},{0}".format(2))
        self.assertEqual('foo bar baz ',"{0:4}{1:4}{2:4}".format("foo","bar","baz"))        
```

Its odd how real python adds space in the right for strings, but the left for numbers.

**Comment by gerbal** [comment]
That's an odd behavior. and not one specified in PEP 3101. I added the spacing tests. Now, I have more more work to do.

**Comment by bnmnetp** [comment]
I commented out the tests not covered by PEP 3101, and accepted this PR.

I would rather have a mostly working version of format for us to use and improve upon than to have this languish as an unaccepted PR.

Brad

**Pull request merged**